yellow-paper/docs/addresses-and-keys/addresses-and-keys.md

```diff
@@ -0,0 +1,903 @@
+- No leakage of non-shareable keys.
+  - Given one (or many) shareable keys, a 3rd part MUST NOT be able to derive any other of a user's keys; be they shareable or non-shareable.
+    - Further, they must not be able to derive any relationships _between_ other keys. (This point is inspired by an attack I found where using the same randomness to derive shareable incoming and outgoing keys, resulted in a 3rd party being able to deduce the _difference_ between master secret keys). 
+
+Nice to have:
+- Shareable nullifier key.
+  - A user can optionally share a "shareable" nullifier key, which would enable a trusted 3rd party to see _when_ a particular note hash has been nullified, but would not divulge the contents of the note, or the circumstances under which the note was nullified (as such info would only be gleanable with the shareable viewing keys).
+  - Note: essentially, imagine if you were given a nullifier key, and the nullifier is computed as `h(note_hash, nullifier_key)`. Then someone with the nullifier key could brute-force attempt to hash every `note_hash` in the Note Hashes Tree with the `nullifier_key` in search of matching nullifiers in the nullifier tree - at which point they learn "this note hash just got spent by Bob". It seemed like a good idea, but in practice it might not be useful (too little information to draw any meaning from) or efficient (too compute intensive).
+
+
+### Note Discovery
+
+Requirements:
+- [Nice to have]: A "tagging" keypair that enables faster brute-force identification of owned notes.
+  - Note: this is useful for rapid handshake discovery, but it is an optimisation, and has trade-offs.
+- [Nice to have]: The ability to generate a sequence of tags between Alice and Bob, in line with our latest "Tag Hopping" ideas.
+
+Considerations:
+- There is no enshrined tagging scheme, currently. Whether to enshrine protocol functions (to enable calls to private account contract functions) or to let apps decide on tagging schemes, is an open debate.
+
+#### Tag Hopping
+
+Given that this is our best-known approach, we include some requirements relating to it:
+
+Requirements:
+- A user Bob can non-interactively generate a sequence of tags for some other user Alice, and non-interactively communicate that sequencer of tags to Alice.
+- If a shared secret (that is used for generating a sequence of tags) is leaked, Bob can non-interactively generate and communicate a new sequence of tags to Alice, without requiring Bob nor Alice to rotate their keys.
+  - Note: if the shared secret is leaked through Bob/Alice accidentally leaking one of their keys, then they might need to actually rotate their keys.
+
+### Constraining key derivations
+
+- An app has the ability to constrain the correct encryption and/or note discovery tagging scheme.
+- An app can _choose_ whether or not to constrain the correct encryption and/or note discovery tagging scheme.
```

Comment:
Shouldn't correct encryption and key derivation be always constrained (except maybe for internal incoming data), as a means to guarantee that the recipient can always receive their notes, and prevent griefing?

Reply:
Based on all private applications I've seen to-date (which don't constrain correct encryption), I'd answer "no". Zcash / aztec connect seem to work fine without enforcing this constraint. In simple payment apps, perhaps users always have a vested interest in seeing their funds get delivered. 🤷‍♂️

Reply:
Yeah, it's a good point. There's lots of room for customisability, which can lead to foot guns.
Those interfaces will be fun to design!